### PR TITLE
Added gradient clipping.

### LINKED ISSUE #61: getting NaN loss
I tried two datasets that consists of 10k and 30k images respectively. Can you please tell me if I need to change any hyperparameter?

### PR DESCRIPTION
**Pull request by npielawski**
This PR fixes #61 and #101.

I looked at the maximum of the model gradients during learning and often some gradients can become very big:
image

I added gradient clipping, which is also used by many other repositories of stable diffusion models.

The training loss is more stable, I didn't get the NaN problem with amp=True anymore (cyan is before PR, magenta is after):
image

I think this may also fix the loss divergence issues I encountered, but I haven't tested enough so far. It may be due to the high gradients shooting the model's weights somewhere where gradients are higher which leads to divergence. I saw someone else having [the same issue here](https://github.com/lucidrains/denoising-diffusion-pytorch/issues/4#issuecomment-860757983).

### CODE REVIEW

**Comment by lucidrains** [comment]
@npielawski thank you!
